File: 700_search_in_a_binary_search_tree/search_bst.js
// JS: https://leetcode.com/problems/search-in-a-binary-search-tree/submissions/1187050135/
// TS: https://leetcode.com/problems/search-in-a-binary-search-tree/submissions/1187051756/
class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

//      4
//     / \
//    2   7
//   / \
//  1   3
let root = new Node(4);
root.left = new Node(2);
root.right = new Node(7);
root.left.left = new Node(1);
root.left.right = new Node(3);

function search_bst(root, val) {
    let results = [];

    function recursion(root) {
        if(!root) return null;
        let current = root;
        if(current.val === val) {
            results.push(current.val);
            if(current.left) {
                results.push(current.left.val);
                recursion(current.left);
            }
            if(current.right) {
                results.push(current.right.val);
                recursion(current.right);
            }

        }
        if(root.left) {
            recursion(root.left);
        }
        if(root.right) {
            recursion(root.right);
        }
    }

    recursion(root);

    return results;
}

function search_bst_LC(root, val) {
    function recursion(rootNode) {
        if(!rootNode) return null;
        if(rootNode.val > val) {
            return recursion(root.left);
        } else if (rootNode.val < val) {
            return recursion(root.right);
        } else {
            return rootNode;
        }
    }

    return recursion(root);
}

console.log(search_bst(root, 2));

File: 700_search_in_a_binary_search_tree/search_bst.py
# Py3: https://leetcode.com/problems/search-in-a-binary-search-tree/submissions/1187019797/
from typing import Optional

class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right  = None

#      4
#     / \
#    2   7
#   / \
#  1   3
root = Node(4)
root.left = Node(2)
root.right = Node(7)
root.left.left = Node(1)
root.left.right = Node(3)

def find_target(root: Optional[Node], val: int) -> Optional[Node]:
    result = []

    def recursion(root):
        current = root
        if current.val == val:
            result.append(current.val)
            if current.left:
                result.append(current.left.val)
            if current.right:
                result.append(current.right.val)
        elif current.left:
            recursion(current.left)
        elif current.right:
            recursion(current.right)
    
    recursion(root)

    return result

# Leetcode solution
def find_target_LC(root: Optional[Node], val: int) -> Optional[Node]:
    def recursion(root):
        if not root:
            return None
        
        if root.val > val:
            return recursion(root.left)
        elif root.val < val:
            return recursion(root.right)
        else:
            return root.val
    
    return recursion(root)

print(find_target(root, 2))
print(find_target(root, 5))
print(find_target_LC(root, 2))
print(find_target_LC(root, 5))